### ISSUE TITLE
Remove twisted.internet._posixstdio.StandardIO.closeStdin

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#6785]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)| @thijstriemstra reported|
|-|-|
|Trac ID|trac#6785|
|Type|task|
|Created|2013-10-10 18:51:54Z|
|Branch|https://github.com/twisted/twisted/tree/remove-closeStdin-6785-2|

[twisted/internet/_posixstdio.py](twisted/internet/_posixstdio.py) contains `closeStdin`, `stopReading` and `startReading` are marked as "Stupid compatibility". All were introduced in [#10420](https://github.com/twisted/twisted/issues/10420)/r15825, 8 years ago. Only one is actually deprecated, or issuing a warning, (`closeStdin`) so the other 2 probably need a deprecation warning before we can remove them..

Attachments:

* [6785_deprecate_warning.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0b0/0b005251a895d7d7711840a6885072c39fe30e28/275ba44604b6dc2e3dfb8669d0187627e66c1622.patch) (790 bytes) - added by icchasethi on 2013-10-15 01:14:51Z - 
* [6785_deprecate_warning.2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0b0/0b005251a895d7d7711840a6885072c39fe30e28/4543096a359919d6ab2bced3d2515089fe8d803f.patch) (768 bytes) - added by icchasethi on 2013-10-15 03:38:05Z - 
* [6785_remove_deprecated_fn.3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0b0/0b005251a895d7d7711840a6885072c39fe30e28/c5c5972ec17e14390bc10af3ff125c259fb011ab.patch) (1052 bytes) - added by icchasethi on 2013-10-20 17:52:03Z - 
* [6785_remove_deprecated_fn.4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0b0/0b005251a895d7d7711840a6885072c39fe30e28/bf519e12e5ba61be9706f3359821da606b1cfb21.patch) (1677 bytes) - added by icchasethi on 2013-11-11 00:53:53Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__6785 6785
type__task task
reporter__thijs thijs
priority__low low
milestone__ 
branch__branches_remove_closeStdin_6785_2 branches/remove-closeStdin-6785-2
branch_author__icchasethi__rwall icchasethi, rwall
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1381431114000000 1381431114000000
changetime__1385912201000000 1385912201000000
version__None None
owner__rwall rwall
cc__icchasethi
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_1"></a>|@thijstriemstra commented|
|-|-|

updating rev nr.

**Comment by twisted-trac** [comment]
|<img alt="icchasethi's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|icchasethi set owner to icchasethi|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_7"></a>|@habnabit commented|
|-|-|

(In [40343]) Branching to 'remove-closeStdin-6785'

**Comment by twisted-trac** [comment]
|[<img alt="habnabit's avatar" src="https://avatars.githubusercontent.com/u/21168?s=50" width="50" height="50">](https://github.com/habnabit)<a name="note_8"></a>|@habnabit commented|
|-|-|

(In [40345]) Applying patch from icchasethi. refs [#6785](https://github.com/twisted/twisted/issues/6785)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_9"></a>|@tomprince commented|
|-|-|

(In [40352]) Apply 6785_deprecate_warning.2.patch from icchasethi.

Refs: [#6785](https://github.com/twisted/twisted/issues/6785)

**Comment by twisted-trac** [comment]
|<img alt="icchasethi's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_10"></a>|icchasethi commented|
|-|-|

[Build Results](https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/remove-closeStdin-6785)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

Thanks for the patch, icchasethi.

The changes look good as far as they go.  A couple more things though:

  1. These new deprecations should probably be attached to a new ticket instead.  This ticket is about removing the already-deprecated `closeStdin`.
  1. All code needs to have unit tests.  Please add unit tests for the new deprecations when you work on that new ticket. :)
  1. Please include a [news fragment](https://twistedmatrix.com/trac/wiki/ReviewProcess#Newsfiles) - in this case, a `.removal` would be appropriate.

Thanks again.

**Comment by twisted-trac** [comment]
|<img alt="icchasethi's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_12"></a>|icchasethi commented|
|-|-|

Thanks for the feedback. Attached patch with topfiles and just removing the function

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_13"></a>|@wallrj commented|
|-|-|

(In [40443]) Branching to 'remove-closeStdin-6785-2'

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_14"></a>|@wallrj commented|
|-|-|

(In [40444]) Apply 6785_remove_deprecated_fn.3.patch  from iccha. Refs [#6785](https://github.com/twisted/twisted/issues/6785)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_15"></a>|@wallrj commented|
|-|-|

Hi iccha,

Thanks for the latest patch.

I created a new branch. I guessed that you want to abandon the previous branch,
which contained the new deprecations.

Notes:
 * All tests still pass: https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/remove-closeStdin-6785-2

Points:
 1. Pyflakes: http://buildbot.twistedmatrix.com/builders/pyflakes/builds/1147/steps/pyflakes/logs/new%20pyflakes%20errors
    1. Unused import warning
 2. twisted/conch/scripts/cftp.py
    1. https://buildbot.twistedmatrix.com/builds/twisted-coverage.py/twisted-coverage.py-r40444/twisted_conch_scripts_cftp.html#n812
    2. I think this is a lingering call to closeStdin which isn't tested.
    3. I think you'll have to change that to use loseWriteConnection instead.
    4. You'll also have to add a test for that function.
 3. twisted/topfiles/6785.removal
    1. I think this would sound better as "twisted.internet._posixstdio.closeStdin, deprecated since Twisted 2.1, has been removed."
    2. I'm guessing the version number, but 2.1 was about 8 years ago judging by
       the modified dates here:
       https://pypi.python.org/packages/source/T/Twisted/
    3. See https://twistedmatrix.com/Releases/Twisted/13.1/NEWS.txt for other removal examples.
 4. Remember to create new tickets for the deprecation of stopReading and startReading; one ticket for each.
    1. Here's a good example of how to deprecate (and test the deprecation of)
       functions and methods using twisted deprecation tools. r31056

Please address or answer the numbered points above and submit a patch against the latest branch.

Thanks.

-RichardW.

**Comment by twisted-trac** [comment]
|<img alt="icchasethi's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_16"></a>|icchasethi commented|
|-|-|

Richard,

Thanks for your feedback. Made changes as per your suggestions. Created ticket for the other fn changes.

How ever was not sure how to test the change in conch script. Any pointers?

Thanks,
Iccha

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_17"></a>|@wallrj commented|
|-|-|

(In [40776]) Apply 6785_remove_deprecated_fn.4.patch from icchasethi. Refs [#6785](https://github.com/twisted/twisted/issues/6785)

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)|@wallrj set owner to @wallrj<br>@wallrj set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_19"></a>|@wallrj set status to `new`|
|-|-|

Replying to [icchasethi](#note_16):
> Thanks for your feedback. Made changes as per your suggestions. Created ticket for the other fn changes.
> How ever was not sure how to test the change in conch script. Any pointers?

Hi Iccha,

This looks mostly fine. A few notes below.

Points:
 1. New patch was against trunk, but there was already a branch for this
    ticket. Remember to create patches against the correct branch.
    1. I manually fixed the conflicts.
 2. You say you created a ticket, but I couldn't find it. Are you sure you saved
    the ticket?
    1. In any case, please add a comment to this ticket, with links to the new
       tickets.
 3. I found a corresponding test for conch.scripts.conch
    1. twisted.conch.test.test_conch.StdioInteractingSessionTests.test_eofReceived
    2. I'll duplicate that testcase for cftp and run buildbot on it.
    3. Ideally, we'd find some way of sharing that code, but maybe that's best
       left for another ticket. There seems to be lots of duplication between
       conch.scripts.conch and conch.scripts.cftp

Please create / link those new tickets. I'll see if I can take care of the
missing unittest. Then I'll run another build and put it back up for a quick
final review.

Thanks.

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_20"></a>|@wallrj removed owner|
|-|-|

Ready for another review in log:branches/remove-closeStdin-6785-2

 * Copied twisted.conch.test.test_conch.StdioInteractingSessionTests to twisted.conch.test.test_cftp.SSHSessionTests
   * Looks like those tests were added in  [#4047](https://github.com/twisted/twisted/issues/4047)
   * Not sure why it's call StdioInteractingSessionTests there. I decided to go
     with SSHSessionTests.
   * It's now great to just copy that code, but presumably there are differences
     between conch.SSHSession and cftp.SSHSession which are currently untested.
   * I don't really want to get into re-factoring Conch tests here
     though. I'll open a new ticket for that if it needs doing.

Build results:
 * https://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/remove-closeStdin-6785-2&num_builds=10

-RichardW.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph set owner to @wallrj|
|-|-|

Thanks icchasethi for your contribution, and thanks rwall for moving this along.

A couple of points to address:

  1. The misc file refers to `twisted.internet._posixstdio.closeStdin`, but:
    1. This a private API (see the underscore?) and therefore by definition of no interest to other users and should not be included in a news file :-).
    2. That's not actually the fully-qualified name of the API, since you skipped the class ;-).

  A better way to refer to this API would be to say something like "twisted.internet.stdio.StandardIO.closeStdin, an alias for loseWriteConnection only available on POSIX and deprecated since 2.1, has been removed".
  2. The proliferation of new fake stuff is somewhat concerning; in this case, `FakeStdio`.  It makes tests harder to maintain.  Especially problematic is that the fake exists in the conch tests, but it's a fake of something that exists in twisted.internet.  Also, an identical class already exists in `twisted.conch.test.test_conch.StdioInteractingSessionTests`.  Longer term it might be nice to move this over to `proto_helpers`, since, for example, our tests would not output warnings or fail if the test fake were to switch to having a `closeStdin` method although the real implementation did not.

Addressing point 2 would be a lot easier if there were a sensible interface that `StandardIO` could implement which users could test for.  You may also be interested in [#3597](https://github.com/twisted/twisted/issues/3597) for cleaning up the larger situation here.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_22"></a>|@glyph commented|
|-|-|

Oops, let me make the review status clear; I'm OK with landing this as-is, except for the news file change.  Please address point 2 if you can, at least by consolidating those two implementations, but add a comment to [#3597](https://github.com/twisted/twisted/issues/3597) indicating that this is one of the areas clean-up is required.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_23"></a>|@wallrj set status to `assigned`|
|-|-|

Replying to [glyph](#note_22):
> Oops, let me make the review status clear; I'm OK with landing this as-is, except for the news file change.  Please address point 2 if you can, at least by consolidating those two implementations, but add a comment to [#3597](https://github.com/twisted/twisted/issues/3597) indicating that this is one of the areas clean-up is required.


Ok. I updated the newsfile and added a single FakeStdio in test_conch and imported that from test_cftp.

I know we're not supposed to import stuff from test modules any more, but I didn't think it was worth cluttering proto_helpers with FakeStdio until it's a more complete and generally useful fake.

Hope that's ok. Builds all pass so I'll merge this later unless I hear anything to the contrary.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_24"></a>|@glyph commented|
|-|-|

Sure, that sounds like a pretty reasonable thing to do.

The important thing w.r.t. test modules is not to import things from them _outside_ Twisted.  You're right that this fake is not worth elevating to the level of "thing users should import" at this point, but we can import it _internally_ just fine; that means there are fewer implementations to go delete and fewer variances in behavior to deal with if someone _does_ decide to make the time to polish it into something release-quality.

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_25"></a>|@wallrj set status to `closed`|
|-|-|

(In [40880]) Merge remove-closeStdin-6785-2

Author: icchasethi, rwall
Reviewers: exarkun, rwall, glyph
Fixes: [#6785](https://github.com/twisted/twisted/issues/6785)

twisted.internet.stdio.StandardIO.closeStdin, an alias for loseWriteConnection
only available on POSIX and deprecated since 2.1, has been removed.
